Add label image output feature to RgbdCamera

kunimatsu-tri opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5498)
<!-- Reviewable:end -->

kunimatsu-tri commented:
+@liangfok can I ask you feature review?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-:-KfGM_3vJjwIeuz8-svx:b-qx06dv)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Yes, I will get to this ASAP.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-:-KfGkU3RSbOXCo6JeGWc:b-1rdfvf)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Checkpoint.

---

Reviewed 3 of 4 files at r1.
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/systems/sensors/test/rgbd_camera_test.cc, line 244 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfK4LLMSpj0TRBFmmhn:-KfK4LLMSpj0TRBFmmho:bmk5lto) ([raw file](https://github.com/robotlocomotion/drake/blob/6a48aa22b1847e1248516e46d5f78aeb35783507/drake/systems/sensors/test/rgbd_camera_test.cc#L244)):*
> ```cpp
>     }
> 
>     EXPECT_EQ(expected_num_colors, actual_colors.size());
> ```

BTW, in addition to counting the number of different colors in the label image, would it be worthwhile to sample a few points on the borders of each object's region to verify that the points in each region are of the same color?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-:-KfK4qLQNO0fitQInIl2:by96gdd)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 4 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[drake/systems/sensors/rgbd_camera.cc, line 49 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfMCLptk9ZABm35OTuy:-KfMCJcKMVi5QFal0qxS:bed7t5n) ([raw file](https://github.com/robotlocomotion/drake/blob/6a48aa22b1847e1248516e46d5f78aeb35783507/drake/systems/sensors/rgbd_camera.cc#L49)):*
> ```cpp
>   double r;
>   double g;
>   double b;
> ```

BTW, consider adding a comment explaining what these variables mean and their valid range?

```c++
// Defines a color based on its three primary additive colors: red, green, and
// blue. Each of these primary additive colors are in the range of [0, 1] where
// 0 is the lowest intensity and 1 is the highest intensity.
struct Color {
  double r;  // red
  double g;  // green
  double b;  // blue
}
```

---

*[drake/systems/sensors/rgbd_camera.cc, line 72 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfMBxp3tAcdwsk0CMIU:-KfMBZuNioyJN4YKA8iI:b6jzvfv) ([raw file](https://github.com/robotlocomotion/drake/blob/6a48aa22b1847e1248516e46d5f78aeb35783507/drake/systems/sensors/rgbd_camera.cc#L72)):*
> ```cpp
> const double kTerrainSize = 100.;
> const Color kTerrainColor{1., 0.898, 0.797};
> const Color kBackgroundColor{0.8, 0.898, 1.};
> ```

BTW, consider adding a comment explaining how these colors were chosen. For example, were they chosen so as to be easily distinguished from the colors in `ColorPalette`?

---

*[drake/systems/sensors/rgbd_camera.cc, line 86 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfMA9EWYZIXanqzzv44:-KfMA9EWYZIXanqzzv45:b-fbqmc2) ([raw file](https://github.com/robotlocomotion/drake/blob/6a48aa22b1847e1248516e46d5f78aeb35783507/drake/systems/sensors/rgbd_camera.cc#L86)):*
> ```cpp
> }
> 
> class ColorPalette {
> ```

BTW, consider adding a brief comment describing what this class does:

```c++
// Creates and holds a palette of colors for visualizing different objects in a
// scene (the intent is for a different color to be applied to each identified
// object). The colors are chosen so as to be easily distinguishable. This color
// palette can hold up to 1536 colors. Black is not part of this color palette.
```

---

*[drake/systems/sensors/rgbd_camera.cc, line 107 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfMBQjr6VfkGbgsBezx:-KfMBQjr6VfkGbgsBezy:b-fk16ns) ([raw file](https://github.com/robotlocomotion/drake/blob/6a48aa22b1847e1248516e46d5f78aeb35783507/drake/systems/sensors/rgbd_camera.cc#L107)):*
> ```cpp
>   }
> 
>   Color& get_color(int index) {
> ```

Should this be `const`?

```c++
const Color& get_color(int index) const {
```

---

*[drake/systems/sensors/rgbd_camera.cc, line 113 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfMEa8wElEQCNECNHLb:-KfMEa8wElEQCNECNHLc:b8mphq2) ([raw file](https://github.com/robotlocomotion/drake/blob/6a48aa22b1847e1248516e46d5f78aeb35783507/drake/systems/sensors/rgbd_camera.cc#L113)):*
> ```cpp
> 
>  private:
>   std::vector<Color> color_palette_;
> ```

BTW, should this be a vector of `const` values?

```
 std::vector<const Color> color_palette_;
```

---

*[drake/systems/sensors/rgbd_camera.h, line 115 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfMF1Mfkz25CAThfiQ0:-KfMF1Mfkz25CAThfiQ1:b-9rrcni) ([raw file](https://github.com/robotlocomotion/drake/blob/6a48aa22b1847e1248516e46d5f78aeb35783507/drake/systems/sensors/rgbd_camera.h#L115)):*
> ```objc
>   /// @param tree The RigidBodyTree containing the geometric description of the
>   /// world. The life span of this parameter must exceed that of this class's
>   /// instance.
> ```

BTW, consider documenting the fact that the maximum number of bodies in the tree must be less than 1536 based on the size of the color pallet used for object labeling. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Completed first pass.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-:-KfMFdGMmx6c67VsrsZA:bl4rucz)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Review status: 1 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/sensors/rgbd_camera.cc, line 49 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfMCLptk9ZABm35OTuy:-KfMaA0YnBmKRguU_C6f:b-6mhyx0) ([raw file](https://github.com/robotlocomotion/drake/blob/6a48aa22b1847e1248516e46d5f78aeb35783507/drake/systems/sensors/rgbd_camera.cc#L49)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding a comment explaining what these variables mean and their valid range?

```c++
// Defines a color based on its three primary additive colors: red, green, and
// blue. Each of these primary additive colors are in the range of [0, 1] where
// 0 is the lowest intensity and 1 is the highest intensity.
struct Color {
  double r;  // red
  double g;  // green
  double b;  // blue
}
```
</blockquote></details>

Great, thanks!

---

*[drake/systems/sensors/rgbd_camera.cc, line 72 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfMBxp3tAcdwsk0CMIU:-KfMiDlzufniEymSbGDY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6a48aa22b1847e1248516e46d5f78aeb35783507/drake/systems/sensors/rgbd_camera.cc#L72)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding a comment explaining how these colors were chosen. For example, were they chosen so as to be easily distinguished from the colors in `ColorPalette`?
</blockquote></details>

Done.

---

*[drake/systems/sensors/rgbd_camera.cc, line 86 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfMA9EWYZIXanqzzv44:-KfMdC7-dl03dywpIDWU:b39u3fw) ([raw file](https://github.com/robotlocomotion/drake/blob/6a48aa22b1847e1248516e46d5f78aeb35783507/drake/systems/sensors/rgbd_camera.cc#L86)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding a brief comment describing what this class does:

```c++
// Creates and holds a palette of colors for visualizing different objects in a
// scene (the intent is for a different color to be applied to each identified
// object). The colors are chosen so as to be easily distinguishable. This color
// palette can hold up to 1536 colors. Black is not part of this color palette.
```
</blockquote></details>

Also great! Thanks!

Done.

---

*[drake/systems/sensors/rgbd_camera.cc, line 107 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfMBQjr6VfkGbgsBezx:-KfMiB-gtNxLkXV7qxlt:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6a48aa22b1847e1248516e46d5f78aeb35783507/drake/systems/sensors/rgbd_camera.cc#L107)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Should this be `const`?

```c++
const Color& get_color(int index) const {
```
</blockquote></details>

Done.

---

*[drake/systems/sensors/rgbd_camera.cc, line 113 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfMEa8wElEQCNECNHLb:-KfMfRVKMzMGEdU76tMp:bl2wvrz) ([raw file](https://github.com/robotlocomotion/drake/blob/6a48aa22b1847e1248516e46d5f78aeb35783507/drake/systems/sensors/rgbd_camera.cc#L113)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, should this be a vector of `const` values?

```
 std::vector<const Color> color_palette_;
```
</blockquote></details>

That code didn't work since `vector::value_type` has to be `CopyAssignable`:
http://stackoverflow.com/questions/8685257/why-cant-you-put-a-const-object-into-a-stl-container

---

*[drake/systems/sensors/rgbd_camera.h, line 115 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfMF1Mfkz25CAThfiQ0:-KfMjaHSk7faexE2wmug:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6a48aa22b1847e1248516e46d5f78aeb35783507/drake/systems/sensors/rgbd_camera.h#L115)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider documenting the fact that the maximum number of bodies in the tree must be less than 1536 based on the size of the color pallet used for object labeling. 
</blockquote></details>

Done.

---

*[drake/systems/sensors/test/rgbd_camera_test.cc, line 244 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfK4LLMSpj0TRBFmmhn:-KfMfcqzuoUmVveld-HD:br9v2zs) ([raw file](https://github.com/robotlocomotion/drake/blob/6a48aa22b1847e1248516e46d5f78aeb35783507/drake/systems/sensors/test/rgbd_camera_test.cc#L244)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, in addition to counting the number of different colors in the label image, would it be worthwhile to sample a few points on the borders of each object's region to verify that the points in each region are of the same color?
</blockquote></details>

will add a check that compares the colors in some sampled points to the ones in the `ColorPalette`. I think the sampling points don't need to be the borders.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Addressed all the comments except adding an unit test. will do asap.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-:-KfMo0A0JEjl7HGEJ7cB:b-lwwuuj)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Added a color checking test for label image. PTAL.

---

Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/systems/sensors/test/rgbd_camera_test.cc, line 244 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfK4LLMSpj0TRBFmmhn:-KfNURRSgzHWD8KdEUrS:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6a48aa22b1847e1248516e46d5f78aeb35783507/drake/systems/sensors/test/rgbd_camera_test.cc#L244)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

will add a check that compares the colors in some sampled points to the ones in the `ColorPalette`. I think the sampling points don't need to be the borders.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-:-KfNUUcHOZzBR3Dtduh4:b2t8yji)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-:-KfNUz9jxoevZNCamw4R:bnfp4nl)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
+@jwnimmer-tri platform review please.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-:-KfNVJzxV8ZcIe7JmPX1:biz0v86)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 4 files at r1, 1 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfNvCvSobMAQ1CNJcwr:-KfNvCvT8EE11hNzB35P:buhiu20) (no related file):*
Please update to resolve merge conflicts.

---

*[drake/systems/sensors/rgbd_camera.h, line 52 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfNuBvZEoyY-SIie5Hb:-KfNuBvZEoyY-SIie5Hc:bh3xh1r) ([raw file](https://github.com/robotlocomotion/drake/blob/8284e56e0a9c800975a6f1a57c83222d3d90df92/drake/systems/sensors/rgbd_camera.h#L52)):*
> ```objc
> ///
> ///   - The label image has three channels in the following order: blue, green
> ///     red. Each channel is represented by a uint8_t.
> ```

From this text, I couldn't figure out the _semantics_ of the label image.  Why is there BGR channels in the label?

Reading the implementation, I see that objects seemed to be assigned unique BGR triples based on their model_instance_id.  Why doesn't the output port simply contain an Image where the pixel values are model_instance_id values?  That seems like what every user would want.  As currently implemented, it doesn't seem like users of this API can even discover the ID, since the id <=> color mapping is private.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 2 files at r5.
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfNvCvSobMAQ1CNJcwr:-KfOKXiyQigmXylRXh5u:b-896fix) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Please update to resolve merge conflicts.
</blockquote></details>

Done.

---

*[drake/systems/sensors/rgbd_camera.h, line 52 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfNuBvZEoyY-SIie5Hb:-KfOKhZgPQxv0g5O3ViZ:bshcyoq) ([raw file](https://github.com/robotlocomotion/drake/blob/8284e56e0a9c800975a6f1a57c83222d3d90df92/drake/systems/sensors/rgbd_camera.h#L52)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

From this text, I couldn't figure out the _semantics_ of the label image.  Why is there BGR channels in the label?

Reading the implementation, I see that objects seemed to be assigned unique BGR triples based on their model_instance_id.  Why doesn't the output port simply contain an Image where the pixel values are model_instance_id values?  That seems like what every user would want.  As currently implemented, it doesn't seem like users of this API can even discover the ID, since the id <=> color mapping is private.
</blockquote></details>

I had a concern about having "model_instance_id image" in mind, but now I noticed that the concern is wrong.  So, I will replace the BGR image with "model_instance_id image".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/sensors/rgbd_camera.h, line 52 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfNuBvZEoyY-SIie5Hb:-KfOMoEBfG38DQGNXZGX:b-ucfknh) ([raw file](https://github.com/robotlocomotion/drake/blob/8284e56e0a9c800975a6f1a57c83222d3d90df92/drake/systems/sensors/rgbd_camera.h#L52)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

I had a concern about having "model_instance_id image" in mind, but now I noticed that the concern is wrong.  So, I will replace the BGR image with "model_instance_id image".
</blockquote></details>

Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/sensors/rgbd_camera.h, line 52 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfNuBvZEoyY-SIie5Hb:-KfOWDwltPYgRmhzjig9:bdusm7p) ([raw file](https://github.com/robotlocomotion/drake/blob/8284e56e0a9c800975a6f1a57c83222d3d90df92/drake/systems/sensors/rgbd_camera.h#L52)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Thanks!
</blockquote></details>

@kunimatsu-tri are you doing anything special to reviewable comment threads like this one?  Maybe like clicking "Acknowledge" or something?  In most cases (in other PRs), when my post is the final one in the discussion, the red "work to be done" bubble would be on your side, and it should should up as gray "one unresolved discussion for me".  But often on your PRs where you respond to the defect but haven't pushed the fix yet, it shows up as a red bubble for me, even though I try to be sure to reply to put the discussion in your queue.  I would like to find a way to not have red bubbles when my work is done, so I am thinking of filing a reviewable defect about this, but I wanted to know if there's something you're doing to cause it before I file that issue?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[drake/systems/sensors/rgbd_camera.h, line 52 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfNuBvZEoyY-SIie5Hb:-KfOXPz9N5gbrEyWyH4K:b9xnxmx) ([raw file](https://github.com/robotlocomotion/drake/blob/8284e56e0a9c800975a6f1a57c83222d3d90df92/drake/systems/sensors/rgbd_camera.h#L52)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

@kunimatsu-tri are you doing anything special to reviewable comment threads like this one?  Maybe like clicking "Acknowledge" or something?  In most cases (in other PRs), when my post is the final one in the discussion, the red "work to be done" bubble would be on your side, and it should should up as gray "one unresolved discussion for me".  But often on your PRs where you respond to the defect but haven't pushed the fix yet, it shows up as a red bubble for me, even though I try to be sure to reply to put the discussion in your queue.  I would like to find a way to not have red bubbles when my work is done, so I am thinking of filing a reviewable defect about this, but I wanted to know if there's something you're doing to cause it before I file that issue?
</blockquote></details>

Ah! I was always clicking "Acknowledge".

Thank you for letting me know. I try not doing that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/sensors/rgbd_camera.h, line 52 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfNuBvZEoyY-SIie5Hb:-KfOYPXwUiiWcs2ZyzBn:bhy8g86) ([raw file](https://github.com/robotlocomotion/drake/blob/8284e56e0a9c800975a6f1a57c83222d3d90df92/drake/systems/sensors/rgbd_camera.h#L52)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Ah! I was always clicking "Acknowledge".

Thank you for letting me know. I try not doing that.
</blockquote></details>

Alright, I'll keep you posted if I see continuing weirdness.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Addressed all the comments. PTAL.

---

Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/sensors/rgbd_camera.h, line 52 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfNuBvZEoyY-SIie5Hb:-KfTkT0rQLssd-AO6gC4:b-8nqj1w) ([raw file](https://github.com/robotlocomotion/drake/blob/8284e56e0a9c800975a6f1a57c83222d3d90df92/drake/systems/sensors/rgbd_camera.h#L52)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Alright, I'll keep you posted if I see continuing weirdness.
</blockquote></details>

Thanks. That helps.

Done for your first comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-:-KfTkbwhDPPmpN6cCJMo:b-y0j4qv)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Thanks.  I will review the rest of the code now.

---

Reviewed 1 of 3 files at r6.
Review status: 2 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-:-KfTlPDYUgZQLoBhCnBO:b8s5m0v)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 3 files at r6.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[drake/systems/sensors/rgbd_camera.cc, line 47 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfTzlb841tQBQmWHFKW:-KfTzlb841tQBQmWHFKX:b34rnbl) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L47)):*
> ```cpp
> namespace {
> 
> const int kPortCameraPose = 3;
> ```

BTW It is a bit odd to have this port index a kConstant, but the other port indices member data within the impl.  Choose one way or the other?  (I actually think this way is better, but either one is acceptable.)

---

*[drake/systems/sensors/rgbd_camera.cc, line 98 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU0yQMFBrDAZuL6daO:-KfU0yQMFBrDAZuL6daP:b-24fhrr) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L98)):*
> ```cpp
> // object). The colors are chosen so as to be easily distinguishable. This color
> // palette can hold up to 1536 colors. Black, white and gray, which has the same
> // value for all the three color channels, are not part of this color palette.
> ```

BTW Saying something like "The intensities are spaced as widely as possible given the number of required colors." would have helped me understand the math in the loop below.

---

*[drake/systems/sensors/rgbd_camera.cc, line 110 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU0lcRV-gzE6rXw2ZT:-KfU0lcRV-gzE6rXw2ZU:b-qe2bx8) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L110)):*
> ```cpp
>       // distinguishable as possible for visualization purpose.  We can add more
>       // colors as needed.
>       const int intensity = 255 - i * 255 / num;
> ```

BTW Its mathematically true, but a `DRAKE_ASSERT(intensity > 0);` is an important invariant and writing it down would highlight it to a reader.

---

*[drake/systems/sensors/rgbd_camera.cc, line 143 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU33b5buH7ohMjKu-0:-KfU33b5buH7ohMjKu-1:btexjcg) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L143)):*
> ```cpp
>                              return (c.r == color.r) &&
>                              (c.g == color.g) &&
>                              (c.b == color.b);
> ```

This duplicates the code for "CompareColors".  Could we just call that from here?

---

*[drake/systems/sensors/rgbd_camera.cc, line 144 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU3JRG2DSb-_FuacgF:-KfU3JRHMOaiaYbm1u8L:b563ibq) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L144)):*
> ```cpp
>                              (c.g == color.g) &&
>                              (c.b == color.b);
>                            });
> ```

FYI This is a linear search for every lookup.  It would be pretty easy to make it constant-time with an `unordered_map`, if you wanted to.  I guess if you have only a few objects in the scene it doesn't matter (because you will only have 6 colors to search), but with the full 1000+ objects in a scene, this likely is expensive.

---

*[drake/systems/sensors/rgbd_camera.cc, line 161 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU2AvJNMJDgf3bunaG:-KfU2AvJNMJDgf3bunaH:bl4y4ki) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L161)):*
> ```cpp
>     normalized.r = static_cast<double>(color.r / 255.);
>     normalized.g = static_cast<double>(color.g / 255.);
>     normalized.b = static_cast<double>(color.b / 255.);
> ```

BTW The `static_cast`s around the whole expression are unnecessary and almost confusing -- if the expression isn't double-valued by the time we evaluate the cast, we've already failed!  It will always be a no-op.  If you want to have the cast for clarity, put it around `(color.r)` and not the division.  I think `/ 255.0` would also be plenty clear without a cast.

---

*[drake/systems/sensors/rgbd_camera.cc, line 167 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU2ZEZdVGpF4-b-kSi:-KfU2ZEZdVGpF4-b-kSj:b3ou1u5) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L167)):*
> ```cpp
>   static bool CompareColors(const Color& lhs, const Color& rhs) {
>     return lhs.r == rhs.r && lhs.g == rhs.g && lhs.b == rhs.b;
>   }
> ```

FYI It would be fine to put this `operator==` on `struct Color` directly, if you wanted to.  (Maybe that would make the `find_if` above simpler also?)

---

*[drake/systems/sensors/rgbd_camera.cc, line 170 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU1U5wOJRUZwlI3lFE:-KfU1U5wOJRUZwlI3lFF:b-w216tm) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L170)):*
> ```cpp
> 
>   // These colors are chosen so as to be easily distinguished from the colors in
>   // color_palette_
> ```

BTW nit Missing period at end of sentence.

---

*[drake/systems/sensors/rgbd_camera.cc, line 170 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU3gPpxaQpaazDedzE:-KfU3gPpxaQpaazDedzF:b-eoilkx) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L170)):*
> ```cpp
> 
>   // These colors are chosen so as to be easily distinguished from the colors in
>   // color_palette_
> ```

BTW Adding "They are guaranteed to be distinct from color_palette_ because none of their intensity elements are identical." here would be informative.

---

*[drake/systems/sensors/rgbd_camera.cc, line 352 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU4RN-66-OCfFMQUzG:-KfU4RN0hgOLvGM_sm87:bvby0oc) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L352)):*
> ```cpp
>   label_buffer_->ReadFrontBufferOff();
>   label_buffer_->Update();
> }
> ```

BTW A bunch of the above code is duplicated for color / depth / label.  Please at least add a TODO to reduce the amount of copy-pasted code.

---

*[drake/systems/sensors/rgbd_camera.cc, line 534 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU4mEvrNcP8czM4NGs:-KfU4mEvrNcP8czM4NGt:bxuuof1) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L534)):*
> ```cpp
>         body->get_model_instance_id());
>     actor_for_label->SetUserTransform(vtk_transform);
>   }
> ```

BTW More duplicated code to remark on in a TODO.

---

*[drake/systems/sensors/rgbd_camera.cc, line 554 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU4oQqhj2IZSg0acqt:-KfU4oQrxmVFznkne8ei:bxuuof1) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L554)):*
> ```cpp
>   label_buffer_->Modified();
>   label_buffer_->Update();
> }
> ```

BTW More duplicated code to remark on in a TODO.

---

*[drake/systems/sensors/rgbd_camera.cc, line 627 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU50wpz-drN6cGH5zM:-KfU50wpz-drN6cGH5zN:bqbad8n) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L627)):*
> ```cpp
> 
>       label_image.at(u, height_reversed)[0] =
>           static_cast<uint16_t>(color_palette_.LookUpId(color));
> ```

This appears to depend on a conversion from u8 palette values, into a double-valued color to vtk, through a renderer, then back as u8 color bytes.  Please explain in a comment why this is robust (or at least that we don't know why, with a TODO, if that's the case).  I know the lighting settings are part of it, but the u8->double->gpu->u8 pipe is less clear.

---

*[drake/systems/sensors/rgbd_camera.h, line 51 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU5fc8VHtHAqnozArY:-KfU5fc9N8Z01nZImLF-:b-do8xmt) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.h#L51)):*
> ```objc
> ///     distance from the sensor origin to the object's surface.
> ///
> ///   - The label image has single channel represented by a uint16_t. The value
> ```

Per cppguide, the use of unsigned here seems to be a defect.  This is not a bitfield or something with modular arithmetic.  It's simply conserving cache footprint by its knowledge that we have only 1538 IDs, so we can represent it in 16 bits.  I think `int16_t` serves the same purpose, and stays within the rules.

---

*[drake/systems/sensors/test/rgbd_camera_test.cc, line 155 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU6M-l71sWBN-JtW73:-KfU6M-mg_SKyUFOPRKg:b-u6nq8o) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/test/rgbd_camera_test.cc#L155)):*
> ```cpp
> }
> 
> const std::array<uint8_t, 4> kSkyColor{{255u, 229u, 204u, 255u}};
> ```

This appears to be unused in the unit test?

---

*[drake/systems/sensors/test/rgbd_camera_test.cc, line 156 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU6tmpBsiVPm9mz-7T:-KfU6tmpBsiVPm9mz-7U:b-ftw5y) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/test/rgbd_camera_test.cc#L156)):*
> ```cpp
> 
> const std::array<uint8_t, 4> kSkyColor{{255u, 229u, 204u, 255u}};
> const std::array<uint8_t, 4> kTerrainColor{{204u, 229u, 255u, 255u}};
> ```

This appears to be required to match the camera's internal constant?  If so, that is brittle.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Addressed all the comments except two of them (will do later). PTAL.

---

Review status: 1 of 4 files reviewed at latest revision, 16 unresolved discussions.

---

*[drake/systems/sensors/rgbd_camera.cc, line 47 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfTzlb841tQBQmWHFKW:-KfWu6PUCTqJmf09ktxI:b-1fu15x) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L47)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW It is a bit odd to have this port index a kConstant, but the other port indices member data within the impl.  Choose one way or the other?  (I actually think this way is better, but either one is acceptable.)
</blockquote></details>

Done.

(I first thought that having the port indices as member variables would be a good idea, but it becomes annoying as the number of ports increase.) 

---

*[drake/systems/sensors/rgbd_camera.cc, line 98 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU0yQMFBrDAZuL6daO:-KfWXZV8VaF3XR2ipT2M:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L98)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Saying something like "The intensities are spaced as widely as possible given the number of required colors." would have helped me understand the math in the loop below.
</blockquote></details>

Done.

---

*[drake/systems/sensors/rgbd_camera.cc, line 110 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU0lcRV-gzE6rXw2ZT:-KfWXmPlrpR4ULRq35pp:b-kblpdj) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L110)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Its mathematically true, but a `DRAKE_ASSERT(intensity > 0);` is an important invariant and writing it down would highlight it to a reader.
</blockquote></details>

Oh that's right. Done.

---

*[drake/systems/sensors/rgbd_camera.cc, line 143 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU33b5buH7ohMjKu-0:-KfWaeX8LQa8X0jkFmc9:b-c8tbsa) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L143)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This duplicates the code for "CompareColors".  Could we just call that from here?
</blockquote></details>

Done with defining`operator==` in `struct Color`.

---

*[drake/systems/sensors/rgbd_camera.cc, line 144 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU3JRG2DSb-_FuacgF:-KfWsEm2OYSEvcEYoTKr:b-cagw15) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L144)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI This is a linear search for every lookup.  It would be pretty easy to make it constant-time with an `unordered_map`, if you wanted to.  I guess if you have only a few objects in the scene it doesn't matter (because you will only have 6 colors to search), but with the full 1000+ objects in a scene, this likely is expensive.
</blockquote></details>

Thanks! will follow your advice.

---

*[drake/systems/sensors/rgbd_camera.cc, line 161 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU2AvJNMJDgf3bunaG:-KfWaBk6sM00vF6udpHx:b-f01ea6) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L161)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW The `static_cast`s around the whole expression are unnecessary and almost confusing -- if the expression isn't double-valued by the time we evaluate the cast, we've already failed!  It will always be a no-op.  If you want to have the cast for clarity, put it around `(color.r)` and not the division.  I think `/ 255.0` would also be plenty clear without a cast.
</blockquote></details>

OMG. My brain must have gotten tired already at this time..

Removed `static_cast`.

---

*[drake/systems/sensors/rgbd_camera.cc, line 167 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU2ZEZdVGpF4-b-kSi:-KfWY5SLZq1mZNWnaGh-:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L167)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI It would be fine to put this `operator==` on `struct Color` directly, if you wanted to.  (Maybe that would make the `find_if` above simpler also?)
</blockquote></details>

Done.

---

*[drake/systems/sensors/rgbd_camera.cc, line 170 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU1U5wOJRUZwlI3lFE:-KfWg0DILLf-O8JJ7O5P:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L170)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW nit Missing period at end of sentence.
</blockquote></details>

Done.

---

*[drake/systems/sensors/rgbd_camera.cc, line 170 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU3gPpxaQpaazDedzE:-KfWgzrGruII1wVKp9FV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L170)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Adding "They are guaranteed to be distinct from color_palette_ because none of their intensity elements are identical." here would be informative.
</blockquote></details>

Done.

---

*[drake/systems/sensors/rgbd_camera.cc, line 352 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU4RN-66-OCfFMQUzG:-KfWkLG96R9yIWHfRiNx:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L352)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW A bunch of the above code is duplicated for color / depth / label.  Please at least add a TODO to reduce the amount of copy-pasted code.
</blockquote></details>

Done.

---

*[drake/systems/sensors/rgbd_camera.cc, line 534 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU4mEvrNcP8czM4NGs:-KfWl2UGFrtU1Y2KHhPD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L534)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW More duplicated code to remark on in a TODO.
</blockquote></details>

Done.

---

*[drake/systems/sensors/rgbd_camera.cc, line 554 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU4oQqhj2IZSg0acqt:-KfWkTVGuqYWi64mbPm9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L554)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW More duplicated code to remark on in a TODO.
</blockquote></details>

Done.

---

*[drake/systems/sensors/rgbd_camera.cc, line 627 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU50wpz-drN6cGH5zM:-KfWrdVMi3iqf0xu8h8s:bnf6ntr) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L627)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This appears to depend on a conversion from u8 palette values, into a double-valued color to vtk, through a renderer, then back as u8 color bytes.  Please explain in a comment why this is robust (or at least that we don't know why, with a TODO, if that's the case).  I know the lighting settings are part of it, but the u8->double->gpu->u8 pipe is less clear.
</blockquote></details>

Added a comment right below the `for` because this is true for color / depth / label. 

---

*[drake/systems/sensors/rgbd_camera.h, line 51 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU5fc8VHtHAqnozArY:-KfWhEIttowsBm1E1cTr:b-tusqcs) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.h#L51)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Per cppguide, the use of unsigned here seems to be a defect.  This is not a bitfield or something with modular arithmetic.  It's simply conserving cache footprint by its knowledge that we have only 1538 IDs, so we can represent it in 16 bits.  I think `int16_t` serves the same purpose, and stays within the rules.
</blockquote></details>

That's a good point! I wonder why I didn't so.

Done.

---

*[drake/systems/sensors/test/rgbd_camera_test.cc, line 155 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU6M-l71sWBN-JtW73:-KfWijMC85Z01bB5ceZD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/test/rgbd_camera_test.cc#L155)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This appears to be unused in the unit test?
</blockquote></details>

Done.

---

*[drake/systems/sensors/test/rgbd_camera_test.cc, line 156 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU6tmpBsiVPm9mz-7T:-KfWs3UJP9qWXwOIOyGn:bi26rl8) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/test/rgbd_camera_test.cc#L156)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This appears to be required to match the camera's internal constant?  If so, that is brittle.
</blockquote></details>

will change not to do this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-:-KfWv9kmtT7cdCoNPwpI:b-dc627p)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-:-KfWwJF3ESqCmiZphrBM:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Review status: 2 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[drake/systems/sensors/rgbd_camera.cc, line 144 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU3JRG2DSb-_FuacgF:-KfgbRhplihRkOzsCkF0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/rgbd_camera.cc#L144)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Thanks! will follow your advice.
</blockquote></details>

Done.

---

*[drake/systems/sensors/test/rgbd_camera_test.cc, line 156 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-KfU6tmpBsiVPm9mz-7T:-Kfgd16jsn4f9jK9uDsY:b-wp3ovu) ([raw file](https://github.com/robotlocomotion/drake/blob/1e8d98b85a0001cc90f5a9e3bb2ea31e3a97803e/drake/systems/sensors/test/rgbd_camera_test.cc#L156)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

will change not to do this.
</blockquote></details>

I want to write an API that enables users to pass arbitrary terrain color in the future, then I will rewrite the test to compare the provided terrain color and the rendered color.

However, let me leave this as TODO for now since I feel it wiil be an API for the rendering world, not for RgbdCamera.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r8, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Addressed the rest of two comments. PTAL.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498#-:-KfgfK8QIhzCHYByrIoO:b-rx0u1l)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 2 of 2 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5498)*
<!-- Sent from Reviewable.io -->
